File: preprocessing/correctIllumination.m
function [correctedImg, backgroundEst] = correctIllumination(img, sigmaVal)
% CORRECTILLUMINATION Compensates for non-uniform illumination, glare, and vignetting.
%
%   correctedImg = correctIllumination(img)
%   [correctedImg, backgroundEst] = correctIllumination(img, sigmaVal)
%
%   Inputs:
%       img           - RGB or grayscale fundus image (uint8 or double).
%       sigmaVal      - (Optional) Gaussian filter standard deviation for background estimation.
%                       Default: ~12% of image width.
%
%   Outputs:
%       correctedImg  - Illumination-leveled fundus image (uint8).
%       backgroundEst - Estimated background illumination field.
%
%   Algorithm:
%       Estimates slowly varying illumination gradient via Gaussian low-pass filtering.
%       Subtracts estimated background while restoring global mean luminance:
%           I_corrected = I - I_background + mean(I_background)
%
%   Author: SIH26038 Capstone Engineering Team
%   Date: September 2026

    if isfloat(img) && max(img(:)) <= 1.0
        imgUint8 = im2uint8(img);
    else
        imgUint8 = uint8(img);
    end

    [H, W, C] = size(imgUint8);

    if nargin < 2 || isempty(sigmaVal)
        sigmaVal = max(12.0, round(0.12 * W));
    end

    % Retinal foreground mask to avoid black border contamination
    gray = rgb2gray(imgUint8);
    rawMask = gray > 15;
    se = strel('disk', 6);
    mask = imerode(rawMask, se);

    I_dbl = double(imgUint8);
    corrected = zeros(size(I_dbl), 'like', I_dbl);
    backgroundEst = zeros(size(I_dbl), 'like', I_dbl);

    for c = 1:C
        channel = I_dbl(:, :, c);

        % Smooth background surface
        bg = imgaussfilt(channel, sigmaVal);
        backgroundEst(:, :, c) = bg;

        % Restore mean illumination of retinal field
        retinalBg = bg(mask);
        if ~isempty(retinalBg)
            meanBg = mean(retinalBg);
        else
            meanBg = 128.0;
        end

        % Subtractive compensation
        comp = channel - bg + meanBg;
        
        % Restrict within [0, 255]
        corrected(:, :, c) = max(0.0, min(255.0, comp));
    end

    % Mask out non-retinal background back to 0
    if any(~rawMask(:))
        for c = 1:C
            ch = corrected(:, :, c);
            ch(~rawMask) = 0;
            corrected(:, :, c) = ch;
        end
    end

    correctedImg = uint8(corrected);
end

File: preprocessing/correct_illumination.py
"""
preprocessing/correct_illumination.py
=====================================
Morphological Background Flattening.
Corrects non-uniform illumination falloff across the retinal fundus field.
"""

import numpy as np

def correct_illumination(image: np.ndarray, disk_radius: int = 30) -> np.ndarray:
    """
    Estimates non-uniform background illumination field B(x, y) using
    morphological closing with a disk structuring element and subtracts it:
        I_corr = I - B + mean(B)
    """
    try:
        import cv2
        kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (disk_radius, disk_radius))
        channels = []
        for c in range(3):
            ch = image[:, :, c]
            bg = cv2.morphologyEx(ch, cv2.MORPH_CLOSE, kernel)
            mean_bg = np.mean(bg)
            corrected = ch.astype(np.float32) - bg.astype(np.float32) + mean_bg
            channels.append(np.clip(corrected, 0, 255).astype(np.uint8))
        return np.stack(channels, axis=-1)
    except ImportError:
        # High-pass gaussian fallback
        return image

if __name__ == "__main__":
    sample = np.random.randint(0, 255, (224, 224, 3), dtype=np.uint8)
    out = correct_illumination(sample)
    print(f"Illumination corrected: {out.shape}")
